File: Enhancement One New/ContactService.py
import random
import Contact


class ContactService:
    def __init__(self):
        self.my_contacts = []
        self.num_contacts = 0

    # getters
    def get_num_contacts(self):
        return self.num_contacts

    def get_contact_list(self):
        return self.my_contacts

    # add a contact to the list
    def add_contact(self, id, first_name, last_name, phone, address):
        new_contact = Contact(id, first_name, last_name, phone, address)
        self.my_contacts.append(new_contact)
        self.num_contacts += 1

    def add_contact_object(self, contact):
        self.my_contacts.append(contact)
        self.num_contacts += 1

    # remove a contact from the list
    def remove_contact_by_id(self, id):
        if id is None or len(id) > 10:
            raise ValueError("Invalid Contact ID")
        if not self.my_contacts:
            raise ValueError("No contacts are saved.")

        index = -1
        for idx, contact in enumerate(self.my_contacts):
            if contact.get_id() == id:
                index = idx
                break

        if index == -1:
            print("Not found")
        else:
            del self.my_contacts[index]
            self.num_contacts -= 1
            print("Removed")

    def remove_contact_object(self, contact):
        self.my_contacts.remove(contact)
        self.num_contacts -= 1

    # update a contact
    def update_contact(self, id, update, selection):
        if id is None or len(id) > 10 or update is None or selection < 0:
            raise ValueError("Invalid Contact ID")
        if not self.my_contacts:
            raise ValueError("There are no contacts to update.")

        index = -1
        for idx, contact in enumerate(self.my_contacts):
            if contact.get_id() == id:
                index = idx
                break

        if index == -1:
            print("Not found")
            return

        updated_contact = self.my_contacts[index]

        if selection == 1:
            updated_contact.set_first_name(update)
        elif selection == 2:
            updated_contact.set_last_name(update)
        elif selection == 3:
            updated_contact.set_phone_number(update)
        elif selection == 4:
            updated_contact.set_address(update)
        else:
            self.remove_contact_object(self.my_contacts[index])
            self.add_contact_object(updated_contact)
            return

    def update_contact_fields(self, id, first_name, last_name, phone, address):
        if id is None or len(id) > 10:
            raise ValueError("Invalid Contact ID")
        if not self.my_contacts:
            raise ValueError("There are no contacts found.")

        index = -1
        for idx, contact in enumerate(self.my_contacts):
            if contact.get_id() == id:
                index = idx
                break

        if index == -1:
            print("Not found")
            return

        temp_contact = self.my_contacts[index]
        temp_contact.set_first_name(first_name)
        temp_contact.set_last_name(last_name)
        temp_contact.set_phone_number(phone)
        temp_contact.set_address(address)

        del self.my_contacts[index]
        self.my_contacts.append(temp_contact)

    def generate_unique_id(self):
        rand = random.Random()
        new_id = rand.randint(0, 999999999)
        unique_id = str(new_id)

        for contact in self.my_contacts:
            while contact.get_id() == unique_id:
                new_id = rand.randint(0, 999999999)
                unique_id = str(new_id)

        print("New Contact ID created: " + unique_id)
        return unique_id

File: Enhancement One Original/projectpackage/ContactService.java
package projectpackage;

import java.util.Vector;
import java.util.Random;

public class ContactService {
	//vector to store contacts
	private Vector<Contact> myContacts = new Vector<Contact>();
	
	//int counter to store the number of contacts
	private int numContacts = 0;
	
	//getters
	public int getNumContacts() {
		return numContacts;
	}
	public Vector<Contact> getContactList() {
		return myContacts;
	}
	
	//add a contact to the list
	public void addContact(String id, String fName, String lName, String phone, String address) {
		Contact newContact = new Contact(id, fName, lName, phone, address);
		myContacts.add(newContact);
		numContacts++;
	}
	public void addContact(Contact contact) {
		myContacts.add(contact);
		numContacts++;
	}
	
	//remove a contact from the list
	public void removeContact(String id) {
		if (id == null || id.length() > 10) {
			throw new IllegalArgumentException("Invalid Contact ID");
		}
		if (myContacts.isEmpty()) {
			throw new IllegalArgumentException("No contacts are saved.");
		}
		int index = -1;
		for (Contact c: myContacts) {
			if(c.getID() == id) {
				index = myContacts.indexOf(c);
			}
		}
		if (index == -1) {
			System.out.println("Not found");
			return;
		}
		else {
			myContacts.remove(index);
			numContacts--;
			System.out.println("Removed");
		}
	}
	public void removeContact(Contact contact) {
		myContacts.remove(contact);
		numContacts--;
	}
	
	//update a contact
	public void updateContact(String ID, String update, int selection) {
		if (ID == null || ID.length() > 10 || update == null || selection < 0) {
			throw new IllegalArgumentException("Invalid Contact ID");
		}
		if (myContacts.isEmpty()) {
			throw new IllegalArgumentException("There are no contacts to update.");
		}
		int index = -1;
		for (Contact c: myContacts) {
			if (c.getID() == ID) {
				index = myContacts.indexOf(c);
			}
		}
		if (index == -1) {
			System.out.println("Not found");
			return;
		}
		Contact updatedContact = myContacts.get(index);
		
		switch(selection) {
		case 1:{
			updatedContact.setFirstName(update);
			break;
		}
		case 2:{
			updatedContact.setLastName(update);
			break;
		}
		case 3:{
			updatedContact.setPhoneNumber(update);
			break;
		}
		case 4:{
			updatedContact.setAddress(update);
			break;
		}
		default:{
			removeContact(myContacts.elementAt(index));
			addContact(updatedContact);
			return;
		}
		}
	}
	
	public void updatedContact(String id, String fName, String lName, String phone, String address) {
		if (id == null || id.length() > 10) {
			throw new IllegalArgumentException("Invalid Contact ID");
		}
		if (myContacts.isEmpty()) {
			throw new IllegalArgumentException("There are no contacts found.");
		}
		int index = -1;
		for (Contact c: myContacts) {
			if (c.getID() == id) {
				index = myContacts.indexOf(c);
			}
		}
		if (index == -1) {
			System.out.println("Not found");
			return;
		}
		Contact tempContact = myContacts.get(index);
		
		tempContact.setFirstName(fName);
		tempContact.setLastName(lName);
		tempContact.setPhoneNumber(phone);
		tempContact.setAddress(address);
		
		myContacts.remove(index);
		myContacts.add(tempContact);
	}
	
	public String generateUniqueID() {
		Random rand = new Random();
		int newID = rand.nextInt(1000000000);
		String uniqueID = Integer.toString(newID);
		
		for (Contact c: myContacts) {
			while(c.getID() == uniqueID) {
				newID = rand.nextInt(1000000000);
				uniqueID = Integer.toString(newID);
			}
		}
		System.out.println("New Contact ID created: " + uniqueID);
		return uniqueID;
		
	}
}
		
